### PR TITLE
S3g sidecar passthru

### PR DESCRIPTION
**Pull request by lukemarsden**
Draft for collaboration/knowledge sharing

### CODE REVIEW

File: src/server/pps/server/s3g_sidecar_proxy.go [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+						req.ContentLength = int64(len(mashed))
+						// prev := req.Header.Get("Content-Length")
+						// req.Header.Set("Content-Length", fmt.Sprintf("%d", len(mashed)))
+						// logrus.Infof("PROXY request switching Content-Length from %d to %d", prev, len(mashed))
+					}
+					// TODO: check whether awsauth.Sign4 reads the whole request
+					// body into memory, if it does that's bad for large writes
+					// and we should figure out how we can stream it to disk
+					// first...
+					awsauth.Sign4(req, awsauth.Credentials{
```
Comment (lukemarsden) [inline src/server/pps/server/s3g_sidecar_proxy.go:177]: Looks like we probably need to plumb the volume with the web identity token into it. See https://pachyderm.slack.com/archives/C01LBA4NSJU/p1666033437705929?thread_ts=1666032276.294569&cid=C01LBA4NSJU

File: goreleaser/docker.yml [hunk 2]
```diff
@@ -190,81 +190,81 @@ dockers:
           - LICENSE
           - licenses
 
-    # arm64 builds follow
```
Comment (lukemarsden) [inline goreleaser/docker.yml:193]: TODO: revert this change before merging, it's just to make it build on my machine

File: examples/spark/s3-out/spark.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+conf.set('spark.hadoop.fs.s3a.path.style.access', 'true')
+conf.set('spark.hadoop.fs.s3a.connection.ssl.enabled', 'false')
+conf.set("spark.hadoop.fs.s3a.change.detection.mode", 'none')
+conf.set("spark.hadoop.fs.s3a.change.detection.version.required", 'false')
+
+sc = SparkContext(conf=conf)
+sc.setLogLevel("ERROR")
+# sc.setLogLevel("DEBUG")
+sc.setSystemProperty("com.amazonaws.services.s3.disablePutObjectMD5Validation", "true")
```
Comment (lukemarsden) [inline examples/spark/s3-out/spark.py:38]: probably don't need this any more

File: examples/spark/s3-out/spark.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
+from pyspark.sql import SparkSession, Row, DataFrame
+from pyspark.context import SparkContext
+from pyspark import SparkConf
+import time
+import os
+
+conf = SparkConf()
+minio = False
```
Comment (lukemarsden) [inline examples/spark/s3-out/spark.py:8]: could strip this out

File: src/server/pps/server/s3g_sidecar_proxy.go [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+}
+
+// TODO: These shouldn't be global variables, it's just expedient for this
+// experimental feature.
+
+// CurrentBucket is the real backend bucket. Set in s3g_sidecar.go
+var CurrentBucket string = "out"
+
+// This is like pipeline_name-<job-id>
+var CurrentTargetPath string = ""
```
Comment (lukemarsden) [inline src/server/pps/server/s3g_sidecar_proxy.go:34]: JobScopedPrefix

File: src/server/pps/server/s3g_sidecar_proxy.go [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+	"runtime/debug"
+	"strings"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/sirupsen/logrus"
+	awsauth "github.com/smartystreets/go-aws-auth"
+)
+
+type RawS3Proxy struct {
```
Comment (lukemarsden) [inline src/server/pps/server/s3g_sidecar_proxy.go:24]: put the global variables inside the struct, duh

**Comment by lukemarsden** [comment]
Protocol: https://docs.google.com/spreadsheets/d/1VwCxYWba9WfmY3oPTKP2x024CYJ1iQ9TCLZ7WMYQoSY/edit#gid=0

**Comment by lukemarsden** [comment]
Wireshark: https://github.com/eldadru/ksniff#ksniff

File: src/server/pps/server/s3g_sidecar_proxy.go [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+							AccessKeyID:     os.Getenv("MINIO_ID"),
+							SecretAccessKey: os.Getenv("MINIO_SECRET"),
+						})
+					}
+				},
+				ModifyResponse: func(resp *http.Response) error {
+					// transform each of the response headers
+					for k, v := range resp.Header {
+						for i, vv := range v {
+							v[i] = transform(vv)
```
Comment (lukemarsden) [inline src/server/pps/server/s3g_sidecar_proxy.go:200]: should be untransform

**Comment by msteffen** [comment]
Closing PR as it's obsolete (though we should consider revisiting the project someday)
